implementations/micrometer-registry-prometheus/src/main/java/io/micrometer/prometheus/PrometheusHistogram.java

```diff
@@ -0,0 +1,146 @@
+import io.micrometer.core.lang.Nullable;
+import io.prometheus.client.exemplars.Exemplar;
+import io.prometheus.client.exemplars.HistogramExemplarSampler;
+
+/**
+ * Internal {@link Histogram} implementation for Prometheus that handles {@link Exemplar exemplars}.
+ *
+ * @author Jonatan Ivanov
+ */
+class PrometheusHistogram extends TimeWindowFixedBoundaryHistogram {
+    private final double[] buckets;
+    private final AtomicReferenceArray<Exemplar> exemplars;
+    @Nullable private final HistogramExemplarSampler exemplarSampler;
+
+    PrometheusHistogram(Clock clock, DistributionStatisticConfig config, @Nullable HistogramExemplarSampler exemplarSampler) {
+        super(
+                clock,
+                DistributionStatisticConfig.builder()
+                        .expiry(Duration.ofDays(1825)) // effectively never roll over
+                        .bufferLength(1)
+                        .build()
+                        .merge(config),
+                true
+        );
+
+        this.exemplarSampler = exemplarSampler;
+        if (isExemplarsEnabled()) {
+            double[] originalBuckets = super.getBuckets();
+            if (originalBuckets[originalBuckets.length - 1] != Double.POSITIVE_INFINITY) {
+                this.buckets = Arrays.copyOf(originalBuckets, originalBuckets.length + 1);
+                this.buckets[buckets.length - 1] = Double.POSITIVE_INFINITY;
+            }
+            else {
+                this.buckets = originalBuckets;
+            }
+            this.exemplars = new AtomicReferenceArray<>(this.buckets.length);
+        }
+        else {
+            this.buckets = null;
+            this.exemplars = null;
+        }
+    }
+
+    boolean isExemplarsEnabled() {
+        return exemplarSampler != null;
+    }
+
+    @Override
+    public void recordDouble(double value) {
+        super.recordDouble(value);
+        if (isExemplarsEnabled()) {
+            updateExemplar(value);
+        }
+    }
+
+    @Override
+    public void recordLong(long value) {
+        super.recordLong(value);
+        if (isExemplarsEnabled()) {
+            updateExemplar(value);
+        }
+    }
+
+    private void updateExemplar(double value) {
+        int index = this.leastLessThanOrEqualTo(value);
+        index = (index == -1) ? exemplars.length() - 1 : index;
+        updateExemplar(value, index);
+    }
+
+    private void updateExemplar(double value, int index) {
+        double bucketFrom = (index == 0) ? Double.NEGATIVE_INFINITY : buckets[index - 1];
+        double bucketTo = buckets[index];
+        Exemplar prev;
+        Exemplar next;
+
+        do {
+            prev = exemplars.get(index);
+            next = exemplarSampler.sample(value, bucketFrom, bucketTo, prev);
+            if (next == null || next == prev) {
+                return;
+            }
+        }
+        while (!exemplars.compareAndSet(index, prev, next));
+    }
+
+    @Nullable Exemplar[] exemplars() {
```

Comment:
It would be great to have a "snapshot" method on `AtomicReferenceArray<T>` that returns `T[]` or `Collection<T>`.

Please let me know if you have a better idea for this.
Using `AtomicReference[]` is somewhat similar: if we don't want to return `AtomicReference[]`, we need to get every element of it and add it to a collection (what we do here).